docs/AGENTS-ui.md

```diff
@@ -0,0 +1,524 @@
+ui/build    # Builds the UI
+ui/run-dev-server    # Starts the development server with live reload
+```
+
+The UI uses:
+
+- **TypeScript** for type safety
+- **Mithril** as the UI framework
+- **Rollup** for bundling
+- **pnpm** for package management
+- **ESLint** for linting (based on Google style)
+- **Playwright** for integration tests
+
+## Plugin Architecture
+
+Plugins are the primary extension mechanism for the UI. They follow this structure:
+
+```typescript
+import {PerfettoPlugin} from '../../public/plugin';
+import {Trace} from '../../public/trace';
+import {App} from '../../public/app';
+
+export default class MyPlugin implements PerfettoPlugin {
+  // Unique reverse-DNS identifier
+  static readonly id = 'com.example.MyPlugin';
+
+  // Optional: Human-readable description
+  static readonly description = 'Does something useful';
+
+  // Optional: Declare dependencies on other plugins
+  static readonly dependencies = [OtherPlugin];
+
+  // Called when the plugin is activated (before trace load)
+  static onActivate(app: App): void {
+    // Register commands, sidebar items, pages that don't need a trace
+  }
+
+  // Called when a trace is loaded
+  async onTraceLoad(trace: Trace): Promise<void> {
+    // Register tracks, tabs, commands that need trace data
+    // Query the trace processor, add tracks to the workspace
+  }
+}
+```
+
+**Plugin Lifecycle:**
+
+1. `onActivate()` - Called when the plugin is enabled, before any trace is loaded. Use for registering global commands, pages, and sidebar items.
+2. `onTraceLoad()` - Called when a trace is loaded. Use for registering tracks, tabs, and commands that depend on trace data.
+3. `trace.onTraceReady` event - Fired after all plugins have finished `onTraceLoad()`. Use for automations that need all tracks to be available.
+
+**Key APIs available to plugins:**
+
+- `trace.engine` - Run SQL queries against TraceProcessor
+- `trace.tracks` - Register and find tracks
+- `trace.selection` - Manage selection state
+- `trace.commands` - Register commands
+- `trace.tabs` - Register tabs in the details panel
+- `trace.timeline` - Access timeline state
+- `trace.workspace` - Manage the track tree structure
+
+## Mithril Patterns and Best Practices
+
+The UI uses Mithril.js. Follow these patterns:
+
+**Component Structure:**
+```typescript
+import m from 'mithril';
+
+interface MyComponentAttrs {
+  readonly value: string;
+  readonly onChange: (newValue: string) => void;
+}
+
+export class MyComponent implements m.ClassComponent<MyComponentAttrs> {
+  // Local state
+  private expanded = false;
+
+  view({attrs}: m.CVnode<MyComponentAttrs>): m.Children {
+    return m('.my-component',
+      m(Button, {label: attrs.value, onclick: () => this.expanded = !this.expanded}),
+      this.expanded && m('.details', 'Expanded content'),
+    );
+  }
+}
+```
+
+**Mithril Rules:**
+
+- No need to call`m.redraw()` from within lifecycle hooks or handlers. Mithril automatically schedules redraws.
```

Comment:
Maybe specify that redraws are triggered automatically:
- in DOM event handlers registered via mithril components
- after trace processor queries complete

Not after manually registered event handlers.